command/cp.go

```diff
@@ -966,3 +1012,57 @@ func guessContentType(file *os.File) string {
 	}
 	return contentType
 }
+
+type CountingWriter struct {
+	pb progressbar.ProgressBar
+	fp *os.File
+}
+
+func (r *CountingWriter) WriteAt(p []byte, off int64) (int, error) {
+	n, err := r.fp.Write(p)
+	if err != nil {
+		return n, err
+	}
+
+	r.pb.AddCompletedBytes(int64(n))
+
+	return n, err
+}
+
+type CountingReader struct {
+	pb      progressbar.ProgressBar
+	fp      *os.File
+	signMap map[int64]struct{}
+	mu      sync.Mutex
+}
+
+func (r *CountingReader) Read(p []byte) (int, error) {
+	n, err := r.fp.Read(p)
+	if err != nil {
+		return n, err
+	}
+	r.pb.AddCompletedBytes(int64(n))
+	return n, err
+}
+
+func (r *CountingReader) ReadAt(p []byte, off int64) (int, error) {
+	n, err := r.fp.ReadAt(p, off)
+	if err != nil {
+		return n, err
+	}
+	r.mu.Lock()
+	// Ignore the first signature call
```

Comment:
Could you please explain in more details why this is necessary?

Reply:
I'd like to know too.

Reply:
The AWS SDK uses double-passing for uploading files. The first one is for computing the `MD5` checksum, and the second one is for uploading the file. It checks that the given byte has been read before. If it is read, it means it is the turn of the upload process.